ckanext/data_qld/resource_freshness/validation.py

```diff
@@ -0,0 +1,53 @@
+# encoding: utf-8
+
+from sqlalchemy.sql.expression import update
+import ckan.plugins.toolkit as tk
+import ckan.lib.navl.dictization_functions as df
+import logging
+import datetime as dt
+import ckan.lib.uploader as uploader
+
+from ckanext.data_qld.resource_freshness.helpers import helpers as h
+
+
+def validate_next_due_date(keys, flattened_data, errors, context):
+    '''
+    Validate the next due date according the `update_frequency` input
+    '''
+    data = df.unflatten(flattened_data)
+    key, = keys
+    if data['update_frequency'] in h.get_update_frequencies():
+        if not data[key]:
+            raise tk.ValidationError({key: [tk._("Missing value")]})
+
+        if dt.datetime.strptime(data[key], '%Y-%m-%d').date() < dt.date.today():
+            raise tk.ValidationError({key: [tk._("Date should be in future")]}) 
+
+        flattened_data[keys] = h.recalculate_due_date(data['update_frequency'], data[key])
+
+
+def validate_nature_of_change_data(keys, flattened_data, errors, context):
+    '''
+    Validate the nature of change data
+    '''
+    data = df.unflatten(flattened_data)
+    res, index, key = keys
+    resource = data[res][index]
+    if not resource.get('id'):
+        return
+
+    file_upload = tk.request.params.get('upload', None)
+    if isinstance(file_upload, uploader.ALLOWED_UPLOAD_TYPES):
+        if not resource[key]:
+            raise tk.ValidationError({key: [tk._("Missing value")]})
+    else:
+        old_resource_id = tk.request.params.get('id')
+        old_res = tk.get_action('resource_show')(context, {'id': old_resource_id})
+        if old_res['url'] == resource['url']:
+            return
+
+    if not resource[key]:
+        raise tk.ValidationError({key: [tk._("Missing value")]})
+
+    if resource[key] == 'add-new-time-series' and data['update_frequency'] in h.get_update_frequencies():
+        flattened_data[('next_update_due',)] = h.recalculate_due_date(data['update_frequency'], data['next_update_due'])
```

Comment:
Use .get(key) for all dictionary objects to minimize the chance of the key not found errors because this new metadata field `next_update_due` might not exist yet. Decide if the default value should return None or empty string depending on how the data is going to be used in the code or downstream methods